File: solutions/_18_4sum/solution.java
package _18_4sum;

import java.util.*;

class Solution {

  public List<List<Integer>> fourSum(int[] nums, int target) {
    List<List<Integer>> result = new ArrayList<>();
    Arrays.sort(nums);
    for (int a = 0; a < nums.length - 3; a++) {
      if (a > 0 && nums[a] == nums[a - 1]) {
        continue;
      }
      for (int b = a + 1; b < nums.length - 2; b++) {
        if (b > a + 1 && nums[b] == nums[b - 1]) {
          continue;
        }

        int c = b + 1;
        int d = nums.length - 1;

        while (c < d) {
          long sum =
            (long) nums[a] + (long) nums[b] + (long) nums[c] + (long) nums[d];
          if (sum == target) {
            result.add(Arrays.asList(nums[a], nums[b], nums[c], nums[d]));
            while (c < d && nums[c] == nums[c + 1]) {
              c++;
            }
            while (c < d && nums[d] == nums[d - 1]) {
              d--;
            }
            c++;
            d--;
          } else if (sum < target) {
            c++;
          } else {
            d--;
          }
        }
      }
    }
    return result;
  }
}
/*
https://leetcode.com/problems/4sum/
*/

File: solutions/_18_4sum/solution.py
from typing import List


class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        length = len(nums)
        result = []

        for a in range(length):
            if a > 0 and nums[a] == nums[a - 1]:
                continue
            for b in range(a + 1, length):
                if b > a + 1 and nums[b] == nums[b - 1]:
                    continue
                c = b + 1
                d = length - 1
                while c < d:
                    sum = nums[a] + nums[b] + nums[c] + nums[d]
                    if sum == target:
                        result.append([nums[a], nums[b], nums[c], nums[d]])
                        c += 1
                        d -= 1
                        while c < d and nums[c] == nums[c - 1]:
                            c += 1
                        while c < d and nums[d] == nums[d + 1]:
                            d -= 1
                    elif sum < target:
                        c += 1
                    else:
                        d -= 1
        return result


# https://leetcode.com/problems/4sum/
